### PR TITLE
deprecated: deprecate array_associate

### PR DESCRIPTION
**Pull request by Goodmain**
Refs #53

### CODE REVIEW

File: src/Iterators/CsvIterator.php [hunk 1]
```diff
@@ -47,12 +47,13 @@ public function current()
...
         }
 
-        return array_associate($this->currentCsvLine, function ($value, $key) {
-            return [
-                'key' => $this->columns[$key],
-                'value' => $value
-            ];
+        $result = [];
+
+        array_walk($this->currentCsvLine, function($value, $key) use (&$result) {
```
Comment (DenTray) [inline src/Iterators/CsvIterator.php:52]: such logic wouldn't remove legacy keys, let's use mapWithKeys helper

```php
collect($this->currentCsvLine)->mapWithKeys(function ($item) {
    return [$item['email'] => $item['name']];
});
```

File: src/Services/HttpRequestService.php [hunk 2]
```diff
@@ -203,11 +203,10 @@ private function setData(string $method, array $headers, array $data = []): void
...
         }
 
-        $lowerHeaders = array_associate($headers, function ($value, $key) {
-            return [
-                'key' => strtolower($key),
-                'value' => $value
-            ];
+        $lowerHeaders = [];
+
+        array_walk($headers, function($value, $key) use (&$lowerHeaders) {
```
Comment (DenTray) [inline src/Services/HttpRequestService.php:208]: as I mentioned above, please use mapWithKeys colection method instead

File: src/helpers.php [hunk 3]
```diff
@@ -256,6 +256,8 @@ function clear_folder($path)
  *  }
  *
  * @return array
+ * 
+ * @deprecated Use array_walk instead
```
Comment (DenTray) [inline src/helpers.php:260]:
```suggestion
 * @deprecated Use array_walk or mapWithKeys collection's method instead
 * https://laravel.com/docs/5.4/collections#method-mapwithkeys
 * https://www.php.net/manual/en/function.array-walk.php
```

File: tests/CsvIteratorTest.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        parent::setUp();
+
+        $this->csvIteratorClass = new CsvIterator($this->getFixturePath('addresses.csv'));
+
+        $this->columnsProperty = new ReflectionProperty(CsvIterator::class, 'columns');
+        $this->columnsProperty->setAccessible(true);
+    }
+
+    public function testOpenFile()
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:23]:
```suggestion
    public function testOpenNotExistsFile()
```

File: tests/CsvIteratorTest.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        $this->expectException(ErrorException::class);
+
+        $this->getMockBuilder(CsvIterator::class)
+            ->setConstructorArgs(['not_exists_file.csv'])
+            ->getMock();
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:29]: could we use regular class creation?

```suggestion
        new CsvIterator($this->getFixturePath('not_exists_file.csv'));
```

File: tests/CsvIteratorTest.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        $this->expectExceptionMessage('fopen(not_exists_file.csv): failed to open stream: No such file or directory');
+    }
+
+    public function testParseColumns()
+    {
+        $header = $this->getJsonFixture('header.json');
+
+        $this->csvIteratorClass->parseColumns($header);
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:38]: let's deprecate the `parseColumns` method and implement `setColumns`

File: tests/CsvIteratorTest.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        $this->assertEquals(0, $rowKey);
+        $this->assertEqualsFixture('current_after_next_line.json', $currentLine);
+    }
+
+    public function testGenerator()
+    {
+        $result = [];
+        $generator = $this->csvIteratorClass->getGenerator();
+
+        foreach ($generator as $columns) {
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:87]:
```suggestion
        foreach ($generator as $row) {
```

File: tests/CsvIteratorTest.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            $result[] = $columns;
+        }
+
+        $rowKey = $this->csvIteratorClass->key();
+
+        $this->assertEquals(6, $rowKey);
+        $this->assertEqualsFixture('all_data.json', $result);
+    }
+
+    public function testGeneratorWithHeader()
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:97]: please also implement test case for `testGeneratorWithHeadersInvalidCount`

**Review by DenTray — CHANGES_REQUESTED** [review]

File: src/Iterators/CsvIterator.php (hunk 1 above)
Comment (Goodmain) [reply src/Iterators/CsvIterator.php:52]: @DenTray There is no legacy keys. getcsv() returns indexed array

File: src/Services/HttpRequestService.php (hunk 2 above)
Comment (Goodmain) [reply src/Services/HttpRequestService.php:208]: @DenTray Same ad above

File: tests/CsvIteratorTest.php (hunk 6 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php:37]: @DenTray Lets move it into separate task with other refactoring related to CsvIterator which I mentioned in our previous discussion

File: tests/CsvIteratorTest.php (hunk 4 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php]: Done

File: tests/CsvIteratorTest.php (hunk 5 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php]: Done

File: tests/CsvIteratorTest.php (hunk 7 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php:86]: Done

File: tests/CsvIteratorTest.php (hunk 8 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php:96]: Done

File: tests/CsvIteratorTest.php [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        $result = [];
+        $header = $this->getJsonFixture('header.json');
+
+        $this->csvIteratorClass->parseColumns($header);
+
+        $generator = $this->csvIteratorClass->getGenerator();
+
+        foreach ($generator as $row) {
+            $result[] = $row;
+        }
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:107]:
```suggestion
        $header = $this->getJsonFixture('header.json');

        $this->csvIteratorClass->parseColumns($header);

        $generator = $this->csvIteratorClass->getGenerator();

        $result = array_map(function ($item) {
            return $item;
        }, $generator);
```

File: tests/CsvIteratorTest.php [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        $this->expectExceptionMessage('Incorrect CSV file');
+    }
+
+    public function testValid()
+    {
+        $this->assertTrue($this->csvIteratorClass->valid());
+
+        $generator = $this->csvIteratorClass->getGenerator();
+
+        foreach ($generator as $columns) {}
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:137]:
```suggestion
        foreach ($generator as $columns) {
        }
```

**Review by DenTray — CHANGES_REQUESTED** [review]

File: src/Iterators/CsvIterator.php (hunk 1 above)
Comment (Goodmain) [reply src/Iterators/CsvIterator.php:52]: @DenTray done with array_combine

File: src/Services/HttpRequestService.php (hunk 2 above)
Comment (Goodmain) [reply src/Services/HttpRequestService.php:208]: @DenTray Done

File: src/helpers.php (hunk 3 above)
Comment (Goodmain) [reply src/helpers.php:260]: @DenTray Done

File: tests/CsvIteratorTest.php (hunk 7 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php:86]: @DenTray Done

File: tests/CsvIteratorTest.php (hunk 9 above)
Comment (Goodmain) [reply tests/CsvIteratorTest.php:107]: @DenTray The second argument for array_map should be an array. So it will not work.

Comment (DenTray) [reply tests/CsvIteratorTest.php:107]: @Goodmain got it

File: tests/CsvIteratorTest.php [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public function testGenerator()
+    {
+        $result = [];
+        $generator = $this->csvIteratorClass->getGenerator();
+
+        foreach ($generator as $columns) {
+            $result[] = $columns;
+        }
```
Comment (DenTray) [inline tests/CsvIteratorTest.php:88]:
```suggestion
        foreach ($generator as $row) {
            $result[] = $row;
        }
```

**Review by DenTray — CHANGES_REQUESTED** [review]

**Review by DenTray — APPROVED** [review]

**Pull request merged**